proposals/0000-extract-runtime.md

```diff
@@ -0,0 +1,419 @@
+Right now we are biased towards mobile platforms only.
+
+### Non-goals
+
+This RFC does **not** aim to:
+
+ - Replace or deprecate proposal [Move iOS and Android Specific Code to Their Own Packages](https://github.com/react-native-community/discussions-and-proposals/pull/49)
+
+## Detailed design
+
+> **NOTE**: Because we are talking about headless code, TurboModules refer to C++ only TurboModules unless
+> otherwise noted.
+
+### Overview
+
+Before proceeding to the description of the suggested changes, let's consider an example of a React Native application 
+with additional OOT platform can be represented in current state (more or less):
+
+![Current State Diagram](./assets/0000-rn-diagram-current-oot.png)
+
+The proposed `react-native-runtime` layer would be everything from React Native that is
+independent of the UI layer (so essentially everything headless) with enough extension points and mount
+points to make it elastic enough to be extended by UI layer or platform layer, if needed.
+
+This includes, but is not limited to:
+
+ - Event Loop
+ - JS engine management
+ - C++-only Turbo Module hosting
+
+And ideally, in the future (out of scope for this RFC):
+ - DevTools interface and debugging host (long term, after the interface gets stabilized)
+ - Handling of Shared and Dedicated Workers
+
+`react-native` would then focus mostly on:
+
+  - Shadow Nodes
+  - Rendering
+  - UI TurboModules (?)
+
+The parts of the React Native that are not stable enough or are just being tested out
+can be kept in `react-native` package until becoming more stable, with less frequent breaking changes
+and then moved to `react-native-runtime` with more stable interface.
+
+---
+
+Having a runtime layer extract some of the pieces to shared layer:
+
+![Diagram with Runtime](./assets/0000-diagram-with-runtime.png)
+
+Now, with React Native layer focusing solely on UI layer, having an additional OOT platform would make the former diagram 
+look more like the following (with additional APIs the user decided to add):
+
+![Diagram with Runtime and OOT](./assets/0000-diagram-runtime-multiplatform.png)
+
+The new layer would take care of `jsi` interface surface and let React Native (core/UI)
+focus on working with UI and executing the user code, no matter the engine is currently selected.
+
+At the heart of `react-native-runtime` would be the event-loop and JS interface, on which other building blocks 
+would be based on.
+
+This would, for example, allow community to implement custom storage solution using `localStorage` API,
+ideally matching the DOM spec. By having the DevTools Interface as part of this package, they can also
+hook up into `Storage` tab from `Dev tools` to provide excellent Developer Experience.
+
+While the `react-native-runtime` would need to maintain those generic interfaces, the React Native community
+can contribute with implementations of such interfaces (JS, EventLoop and DevTools).
+
```

Comment:
I think for this to be included in the core, ideally there are zero built-in turbo modules and all the package does is allowing to hook into it from the user space, so that the community can be responsible for providing the implementations of e.g. `localStorage`. One could argue that the react-native-runtime could expose the interfaces that could be based on WebAPI, but I'm not sure that's necessary and to me ideally it should be optional and possible to include my own interfaces as a user

Reply:
Rephrased the paragraph as it was confusing.